sw/airborne/peripherals/invensense3.c

```diff
@@ -0,0 +1,875 @@
+
+  // Go through the different samples
+  uint8_t i = 0;
+  uint8_t j = 0;
+  int32_t temp = 0;
+  uint16_t gyro_samplerate_count;
+  uint16_t accel_samplerate_count;
+  for(uint8_t sample = 0; sample < samples; sample++) { 
+
+    if ((data[0] & 0xFC) != 0x68) {
+        // no or bad data
+    } else {
+      
+      gyro_samplerate_count = gyro_samplerate * (sample + 1);
+      if(gyro_samplerate_count % faster_odr == 0) {
+        gyro[i].p =  (int16_t)((uint16_t)data[7]  | data[8]  << 8 );
+        gyro[i].q =  (int16_t)((uint16_t)data[9]  | data[10] << 8 );
+        gyro[i].r =  (int16_t)((uint16_t)data[11] | data[12] << 8 );
+        
+        // Temperature sensor register data TEMP_DATA is updated with new data at max(Accelerometer ODR, Gyroscope ODR)
+        if(gyro_samplerate == faster_odr)
+          temp += data[13];
+        
+        i++;
+      }
+
+      accel_samplerate_count = accel_samplerate * (sample + 1);
+      if(accel_samplerate_count % faster_odr == 0) {
+        accel[j].x =  (int16_t)((uint16_t)data[1] | data[2] << 8 );
+        accel[j].y =  (int16_t)((uint16_t)data[3] | data[4] << 8 );
+        accel[j].z =  (int16_t)((uint16_t)data[5] | data[6] << 8 );
+
+        if(accel_samplerate == faster_odr)
+          temp += data[13];
+
+        j++;
+      }
+    }
+
+      data += invensense3_fifo_sample_size[inv->sample_size];
+  }
+
+  // ADC temp * temp sensitivity + temp zero
+  float temp_f = ((float)temp / i) / 2.07 + 25.f; 
+  if (accel_samplerate == faster_odr)
+    temp_f = ((float)temp / j) / 2.07 + 25.f; 
+
+  // Send the scaled values over ABI
+  uint32_t now_ts = get_sys_time_usec();
+  AbiSendMsgIMU_GYRO_RAW(inv->abi_id, now_ts, gyro, i, temp_f);
+  AbiSendMsgIMU_ACCEL_RAW(inv->abi_id, now_ts, accel, j, temp_f);
+  AbiSendMsgTEMPERATURE(inv->abi_id, temp_f);
+}
+
+/**
+ * @brief Parse data from registers
+ * 
+ * @param inv The invensense v3 instance
+ * @param data The data from all registers (DATA_TEMP, DATA_ACCEL and DATA_GYRO)
+ */
+static void invensense3_parse_reg_data(struct invensense3_t *inv, uint8_t *data) {
+  static struct Int32Vect3 accel[1] = {0};
+  static struct Int32Rates gyro[1] = {0};
+  
+  int32_t temp = (int16_t)((uint16_t)data[0]  | data[1] << 8);
+  // ADC temp * temp sensitivity + temp zero
+  float temp_f = (float)temp / 132.48 + 25.f;
+
+  accel[0].x =  (int16_t)((uint16_t)data[2] | data[3] << 8 );
+  accel[0].y =  (int16_t)((uint16_t)data[4] | data[5] << 8 );
+  accel[0].z =  (int16_t)((uint16_t)data[6] | data[7] << 8 );
+
+  gyro[0].p =  (int16_t)((uint16_t)data[8] | data[9]  << 8 );
+  gyro[0].q =  (int16_t)((uint16_t)data[10]| data[11] << 8 );
+  gyro[0].r =  (int16_t)((uint16_t)data[12]| data[13] << 8 );
+
+  // Send the scaled values over ABI
+  uint32_t now_ts = get_sys_time_usec();
+  AbiSendMsgIMU_GYRO_RAW(inv->abi_id, now_ts, gyro, 1, temp_f);
+  AbiSendMsgIMU_ACCEL_RAW(inv->abi_id, now_ts, accel, 1, temp_f);
+  AbiSendMsgTEMPERATURE(inv->abi_id, temp_f);
```

Comment:
I would prefer to keep it in since we have multiple sensors using the same driver on the same device. Else that makes it difficult and a lot of the same code duplicated everywhere. That makes it in my opinion worse. If you look at all the IMU's they are like 99% the same code over and over again, while now if you use this peripheral you don't have that

Reply:
Ok, let's keep it like this. We'll see later if it needs modifications.